// n8n-nodes-base/dist/nodes/Airtop/actions/file/delete.operation.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var delete_operation_exports = {};
__export(delete_operation_exports, {
  description: () => description,
  execute: () => execute
});
module.exports = __toCommonJS(delete_operation_exports);
var import_n8n_workflow = require("n8n-workflow");
var import_constants = require("../../constants");
var import_transport = require("../../transport");
const description = [
  {
    displayName: "File ID",
    name: "fileId",
    type: "string",
    default: "",
    required: true,
    description: "ID of the file to delete",
    displayOptions: {
      show: {
        resource: ["file"],
        operation: ["deleteFile"]
      }
    }
  }
];
async function execute(index) {
  const fileId = this.getNodeParameter("fileId", index, "");
  if (!fileId) {
    throw new import_n8n_workflow.NodeOperationError(
      this.getNode(),
      import_constants.ERROR_MESSAGES.REQUIRED_PARAMETER.replace("{{field}}", "File ID")
    );
  }
  await import_transport.apiRequest.call(this, "DELETE", `/files/${fileId}`);
  return this.helpers.returnJsonArray({ data: { message: "File deleted successfully" } });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  description,
  execute
});
