// n8n-nodes-base/dist/nodes/Airtop/actions/file/File.resource.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var File_resource_exports = {};
__export(File_resource_exports, {
  deleteFile: () => deleteFile,
  description: () => description,
  get: () => get,
  getMany: () => getMany,
  load: () => load,
  upload: () => upload
});
module.exports = __toCommonJS(File_resource_exports);
var deleteFile = __toESM(require("./delete.operation"));
var get = __toESM(require("./get.operation"));
var getMany = __toESM(require("./getMany.operation"));
var load = __toESM(require("./load.operation"));
var upload = __toESM(require("./upload.operation"));
const description = [
  {
    displayName: "Operation",
    name: "operation",
    type: "options",
    noDataExpression: true,
    displayOptions: {
      show: {
        resource: ["file"]
      }
    },
    options: [
      {
        name: "Delete",
        value: "deleteFile",
        description: "Delete an uploaded file",
        action: "Delete a file"
      },
      {
        name: "Get",
        value: "get",
        description: "Get a details of an uploaded file",
        action: "Get a file"
      },
      {
        name: "Get Many",
        value: "getMany",
        description: "Get details of multiple uploaded files",
        action: "Get many files"
      },
      {
        name: "Load",
        value: "load",
        description: "Load a file into a session",
        action: "Load a file"
      },
      {
        name: "Upload",
        value: "upload",
        description: "Upload a file into a session",
        action: "Upload a file"
      }
    ],
    default: "getMany"
  },
  ...deleteFile.description,
  ...get.description,
  ...getMany.description,
  ...load.description,
  ...upload.description
];
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  deleteFile,
  description,
  get,
  getMany,
  load,
  upload
});
