// n8n-nodes-base/dist/credentials/PerplexityApi.credentials.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PerplexityApi_credentials_exports = {};
__export(PerplexityApi_credentials_exports, {
  PerplexityApi: () => PerplexityApi
});
module.exports = __toCommonJS(PerplexityApi_credentials_exports);
class PerplexityApi {
  constructor() {
    this.name = "perplexityApi";
    this.displayName = "Perplexity API";
    this.documentationUrl = "perplexity";
    this.properties = [
      {
        displayName: "API Key",
        name: "apiKey",
        type: "string",
        typeOptions: { password: true },
        required: true,
        default: "",
        description: "Your Perplexity API key. Get it from your Perplexity account."
      }
    ];
    this.authenticate = {
      type: "generic",
      properties: {
        headers: {
          Authorization: "=Bearer {{$credentials.apiKey}}"
        }
      }
    };
    this.test = {
      request: {
        baseURL: "https://api.perplexity.ai",
        url: "/chat/completions",
        method: "POST",
        body: {
          model: "r1-1776",
          messages: [{ role: "user", content: "test" }]
        },
        headers: {
          Authorization: "=Bearer {{$credentials.apiKey}}",
          "Content-Type": "application/json"
        },
        json: true
      }
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PerplexityApi
});
