// n8n-nodes-base/dist/nodes/Airtop/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  BASE_URL: () => BASE_URL,
  DEFAULT_TIMEOUT_MINUTES: () => DEFAULT_TIMEOUT_MINUTES,
  ERROR_MESSAGES: () => ERROR_MESSAGES,
  INTEGRATION_URL: () => INTEGRATION_URL,
  MAX_TIMEOUT_MINUTES: () => MAX_TIMEOUT_MINUTES,
  MIN_TIMEOUT_MINUTES: () => MIN_TIMEOUT_MINUTES,
  N8N_VERSION: () => N8N_VERSION,
  OPERATION_TIMEOUT: () => OPERATION_TIMEOUT,
  SESSION_STATUS: () => SESSION_STATUS,
  getN8NVersion: () => getN8NVersion
});
module.exports = __toCommonJS(constants_exports);
var import_fs = require("fs");
var import_n8n_workflow = require("n8n-workflow");
var import_path = require("path");
const getN8NVersion = () => {
  if (process.env.N8N_VERSION) {
    return process.env.N8N_VERSION;
  }
  try {
    const PACKAGE_DIR = (0, import_path.resolve)(__dirname, "../../../");
    const packageJsonPath = (0, import_path.join)(PACKAGE_DIR, "package.json");
    const n8nPackageJson = (0, import_n8n_workflow.jsonParse)((0, import_fs.readFileSync)(packageJsonPath, "utf8"));
    return n8nPackageJson.version;
  } catch (error) {
    return "0.0.0";
  }
};
const N8N_VERSION = getN8NVersion();
const BASE_URL = process.env.AIRTOP_BASE_URL ?? "https://api.airtop.ai/api/v1";
const INTEGRATION_URL = process.env.AIRTOP_INTEGRATION_URL ?? "https://portal-api.airtop.ai/integrations/v1/no-code";
const DEFAULT_TIMEOUT_MINUTES = 10;
const MIN_TIMEOUT_MINUTES = 1;
const MAX_TIMEOUT_MINUTES = 10080;
const SESSION_STATUS = {
  INITIALIZING: "initializing",
  RUNNING: "running"
};
const OPERATION_TIMEOUT = 5 * 60 * 1e3;
const ERROR_MESSAGES = {
  SESSION_ID_REQUIRED: "Please fill the 'Session ID' parameter",
  WINDOW_ID_REQUIRED: "Please fill the 'Window ID' parameter",
  URL_REQUIRED: "Please fill the 'URL' parameter",
  PROFILE_NAME_INVALID: "'Profile Name' should only contain letters, numbers and dashes",
  TIMEOUT_MINUTES_INVALID: `Timeout must be between ${MIN_TIMEOUT_MINUTES} and ${MAX_TIMEOUT_MINUTES} minutes`,
  TIMEOUT_REACHED: "Timeout reached while waiting for the operation to complete",
  URL_INVALID: "'URL' must start with 'http' or 'https'",
  PROFILE_NAME_REQUIRED: "'Profile Name' is required when 'Save Profile' is enabled",
  REQUIRED_PARAMETER: "Please fill the '{{field}}' parameter",
  PROXY_URL_REQUIRED: "Please fill the 'Proxy URL' parameter",
  PROXY_URL_INVALID: "'Proxy URL' must start with 'http' or 'https'",
  SCREEN_RESOLUTION_INVALID: "'Screen Resolution' must be in the format 'width x height' (e.g. '1280x720')",
  SCROLL_BY_AMOUNT_INVALID: "'Scroll By' amount must be a number and either a percentage or pixels (e.g. '100px' or '100%')",
  SCROLL_MODE_INVALID: "Please fill any of the 'Scroll To Edge' or 'Scroll By' parameters"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BASE_URL,
  DEFAULT_TIMEOUT_MINUTES,
  ERROR_MESSAGES,
  INTEGRATION_URL,
  MAX_TIMEOUT_MINUTES,
  MIN_TIMEOUT_MINUTES,
  N8N_VERSION,
  OPERATION_TIMEOUT,
  SESSION_STATUS,
  getN8NVersion
});
