// n8n-nodes-base/dist/nodes/Perplexity/Perplexity.node.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Perplexity_node_exports = {};
__export(Perplexity_node_exports, {
  Perplexity: () => Perplexity
});
module.exports = __toCommonJS(Perplexity_node_exports);
var import_n8n_workflow = require("n8n-workflow");
var import_descriptions = require("./descriptions");
class Perplexity {
  constructor() {
    this.description = {
      displayName: "Perplexity",
      name: "perplexity",
      icon: {
        light: "file:perplexity.svg",
        dark: "file:perplexity.dark.svg"
      },
      group: ["transform"],
      version: 1,
      subtitle: '={{ $parameter["operation"] + ": " + $parameter["resource"] }}',
      description: "Interact with the Perplexity API to generate AI responses with citations",
      defaults: {
        name: "Perplexity"
      },
      inputs: [import_n8n_workflow.NodeConnectionTypes.Main],
      outputs: [import_n8n_workflow.NodeConnectionTypes.Main],
      usableAsTool: true,
      credentials: [
        {
          name: "perplexityApi",
          required: true
        }
      ],
      requestDefaults: {
        baseURL: "https://api.perplexity.ai",
        ignoreHttpStatusErrors: true
      },
      properties: [
        {
          displayName: "Resource",
          name: "resource",
          type: "hidden",
          noDataExpression: true,
          options: [
            {
              name: "Chat",
              value: "chat"
            }
          ],
          default: "chat"
        },
        ...import_descriptions.chat.description
      ]
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Perplexity
});
