// @n8n/n8n-nodes-langchain/dist/nodes/chains/ChainLLM/methods/processItem.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var processItem_exports = {};
__export(processItem_exports, {
  processItem: () => processItem
});
module.exports = __toCommonJS(processItem_exports);
var import_n8n_workflow = require("n8n-workflow");
var import_helpers = require("../../../../utils/helpers");
var import_N8nOutputParser = require("../../../../utils/output_parsers/N8nOutputParser");
var import_chainExecutor = require("./chainExecutor");
const processItem = async (ctx, itemIndex) => {
  const llm = await ctx.getInputConnectionData(
    import_n8n_workflow.NodeConnectionTypes.AiLanguageModel,
    0
  );
  const outputParser = await (0, import_N8nOutputParser.getOptionalOutputParser)(ctx);
  let prompt;
  if (ctx.getNode().typeVersion <= 1.3) {
    prompt = ctx.getNodeParameter("prompt", itemIndex);
  } else {
    prompt = (0, import_helpers.getPromptInputByType)({
      ctx,
      i: itemIndex,
      inputKey: "text",
      promptTypeKey: "promptType"
    });
  }
  if (prompt === void 0) {
    throw new import_n8n_workflow.NodeOperationError(ctx.getNode(), "The 'prompt' parameter is empty.");
  }
  const messages = ctx.getNodeParameter(
    "messages.messageValues",
    itemIndex,
    []
  );
  return await (0, import_chainExecutor.executeChain)({
    context: ctx,
    itemIndex,
    query: prompt,
    llm,
    outputParser,
    messages
  });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  processItem
});
